common/beerocks/bwl/shared/nl80211_client_impl.cpp

```diff
@@ -17,13 +21,276 @@
+
+    static struct nla_policy rate_policy[NL80211_RATE_INFO_MAX + 1];
+    rate_policy[NL80211_RATE_INFO_BITRATE]   = {NLA_U16, 0, 0};
+    rate_policy[NL80211_RATE_INFO_BITRATE32] = {NLA_U32, 0, 0};
+
+    struct channels_ctx {
+        int last_band;
+        bool width_40;
+        bool width_80;
+        bool width_160;
+    } ctx{};
+
+    radio_info = {};
+
+    if (!m_socket) {
+        LOG(ERROR) << "Socket is NULL!";
+        return false;
+    }
+
+    // Get the interface index for given interface name
+    int iface_index = if_nametoindex(interface_name.c_str());
+    if (0 == iface_index) {
+        LOG(ERROR) << "Failed to read the index of interface " << interface_name << ": "
+                   << strerror(errno);
+
+        return false;
+    }
+
+    return m_socket.get()->send_receive_msg(
+        NL80211_CMD_GET_WIPHY, 0,
+        [&](struct nl_msg *msg) -> bool {
+            nla_put_u32(msg, NL80211_ATTR_IFINDEX, iface_index);
+
+            return true;
+        },
+        [&](struct nl_msg *msg) -> bool {
+            struct nlattr *tb[NL80211_ATTR_MAX + 1];
+            struct genlmsghdr *gnlh = (struct genlmsghdr *)nlmsg_data(nlmsg_hdr(msg));
+
+            // Parse the netlink message
+            if (nla_parse(tb, NL80211_ATTR_MAX, genlmsg_attrdata(gnlh, 0), genlmsg_attrlen(gnlh, 0),
+                          NULL)) {
+                LOG(ERROR) << "Failed to parse netlink message!";
+                return false;
+            }
+
+            if (tb[NL80211_ATTR_WIPHY_BANDS]) {
+                struct nlattr *nl_band;
+                int rem_band;
+
+                nla_for_each_nested(nl_band, tb[NL80211_ATTR_WIPHY_BANDS], rem_band)
+                {
+                    struct nlattr *tb_band[NL80211_BAND_ATTR_MAX + 1];
+                    sBandInfo band{};
+
+                    if (ctx.last_band != nl_band->nla_type) {
+                        ctx.width_40  = false;
+                        ctx.width_80  = false;
+                        ctx.width_160 = false;
+                        ctx.last_band = nl_band->nla_type;
+                    }
+
+                    if (nla_parse(tb_band, NL80211_BAND_ATTR_MAX,
+                                  static_cast<nlattr *>(nla_data(nl_band)), nla_len(nl_band),
+                                  NULL)) {
+                        LOG(ERROR) << "Failed to parse wiphy band " << nl_band->nla_type + 1 << "!";
+                        return false;
+                    }
+
+                    if (tb_band[NL80211_BAND_ATTR_HT_CAPA]) {
+                        band.ht_capability = nla_get_u16(tb_band[NL80211_BAND_ATTR_HT_CAPA]);
+
+                        if (band.ht_capability & BIT(1)) {
+                            ctx.width_40 = true;
+                        }
+                    } else {
+                        LOG(DEBUG) << "NL80211_BAND_ATTR_HT_CAPA attribute is missing";
+                    }
+
+                    if (tb_band[NL80211_BAND_ATTR_HT_MCS_SET]) {
```

Comment:
(nitpick) this is only useful/valid if HT_CAPA was set, and it's an ERROR if HT_CAPA was set and HT_MCS is not set.

Reply:
Changed to use NL80211_BAND_ATTR_HT_MCS_SET only if NL80211_BAND_ATTR_HT_CAPA was set and ignored otherwise